File: sdks/nodejs-agent/examples/test-availability.js
/**
 * Test script for availability search
 * 
 * Usage:
 *   1. Copy .env.example to .env and fill in your credentials
 *   2. Run: node examples/test-availability.js
 * 
 * Or set environment variables:
 *   BASE_URL=http://localhost:8080 JWT_TOKEN=your_token node examples/test-availability.js
 */

// Load environment variables from .env file if available
try {
  require('dotenv').config();
} catch (e) {
  // dotenv not installed, use environment variables directly
}

const { CarHireClient, Config, AvailabilityCriteria } = require('../dist/index.js');

// Get configuration from environment variables
const baseUrl = process.env.BASE_URL || 'http://localhost:8080';
const token = process.env.JWT_TOKEN || '';
const agentId = process.env.AGENT_ID || '';

if (!token) {
  console.error('Error: JWT_TOKEN environment variable is required');
  console.error('Please set JWT_TOKEN or create a .env file with your credentials');
  process.exit(1);
}

// Create configuration
const config = Config.forRest({
  baseUrl,
  token: `Bearer ${token}`,
  agentId: agentId || undefined,
});

// Create client
const client = new CarHireClient(config);

// Test data from environment variables
const pickupLocode = process.env.PICKUP_LOCODE || 'PKKHI';
const returnLocode = process.env.RETURN_LOCODE || 'PKLHE';
const pickupDate = process.env.PICKUP_DATE || '2025-12-01T10:00:00Z';
const returnDate = process.env.RETURN_DATE || '2025-12-03T10:00:00Z';
const driverAge = parseInt(process.env.DRIVER_AGE || '28', 10);
const currency = process.env.CURRENCY || 'USD';
const agreementRef = process.env.AGREEMENT_REF || 'AGR-001';

async function testAvailability() {
  try {
    console.log('=== Testing Availability Search ===');
    console.log(`Base URL: ${baseUrl}`);
    console.log(`Pickup: ${pickupLocode} at ${pickupDate}`);
    console.log(`Return: ${returnLocode} at ${returnDate}`);
    console.log(`Driver Age: ${driverAge}, Currency: ${currency}`);
    console.log(`Agreement: ${agreementRef}`);
    console.log('');

    // Create availability criteria
    const criteria = AvailabilityCriteria.make({
      pickupLocode,
      returnLocode,
      pickupAt: new Date(pickupDate),
      returnAt: new Date(returnDate),
      driverAge,
      currency,
      agreementRefs: [agreementRef],
    });

    console.log('Searching availability...');
    console.log('');

    // Search availability (streaming)
    let chunkCount = 0;
    let totalOffers = 0;

    for await (const chunk of client.getAvailability().search(criteria)) {
      chunkCount++;
      const status = chunk.status || 'PARTIAL';
      const items = chunk.items || [];
      totalOffers += items.length;

      console.log(`[Chunk ${chunkCount}] Status: ${status}, Offers: ${items.length}`);

      if (items.length > 0) {
        // Show first offer as example
        const firstOffer = items[0];
        console.log(`  Example offer: ${firstOffer.vehicle_class || 'N/A'} - ${firstOffer.make_model || 'N/A'}`);
        console.log(`    Price: ${firstOffer.currency || currency} ${firstOffer.total_price || 'N/A'}`);
        console.log(`    Source: ${firstOffer.source_id || 'N/A'}`);
      }

      if (status === 'COMPLETE') {
        console.log('');
        console.log(`✓ Search complete! Total chunks: ${chunkCount}, Total offers: ${totalOffers}`);
        break;
      }
    }

    if (chunkCount === 0) {
      console.log('⚠ No availability chunks received');
    }

  } catch (error) {
    console.error('❌ Error:', error.message);
    if (error.statusCode) {
      console.error(`   Status Code: ${error.statusCode}`);
    }
    if (error.code) {
      console.error(`   Error Code: ${error.code}`);
    }
    process.exit(1);
  }
}

// Run test
testAvailability();

File: sdks/python-agent/examples/test-availability.py
#!/usr/bin/env python3
"""
Test script for availability search

Usage:
    1. Copy .env.example to .env and fill in your credentials
    2. Run: python examples/test-availability.py

Or set environment variables:
    BASE_URL=http://localhost:8080 JWT_TOKEN=your_token python examples/test-availability.py
"""

import os
import sys
from datetime import datetime

# Add parent directory to path to import SDK
sys.path.insert(0, os.path.join(os.path.dirname(__file__), '..'))

try:
    from dotenv import load_dotenv
    load_dotenv()
except ImportError:
    # dotenv not installed, use environment variables directly
    pass

from carhire import CarHireClient, Config, AvailabilityCriteria

# Get configuration from environment variables
base_url = os.getenv('BASE_URL', 'http://localhost:8080')
token = os.getenv('JWT_TOKEN', '')
agent_id = os.getenv('AGENT_ID', '')

if not token:
    print('Error: JWT_TOKEN environment variable is required')
    print('Please set JWT_TOKEN or create a .env file with your credentials')
    sys.exit(1)

# Create configuration
config = Config.for_rest({
    'baseUrl': base_url,
    'token': f'Bearer {token}',
    'agentId': agent_id or None,
})

# Create client
client = CarHireClient(config)

# Test data from environment variables
pickup_locode = os.getenv('PICKUP_LOCODE', 'PKKHI')
return_locode = os.getenv('RETURN_LOCODE', 'PKLHE')
pickup_date = os.getenv('PICKUP_DATE', '2025-12-01T10:00:00Z')
return_date = os.getenv('RETURN_DATE', '2025-12-03T10:00:00Z')
driver_age = int(os.getenv('DRIVER_AGE', '28'))
currency = os.getenv('CURRENCY', 'USD')
agreement_ref = os.getenv('AGREEMENT_REF', 'AGR-001')


async def test_availability():
    try:
        print('=== Testing Availability Search ===')
        print(f'Base URL: {base_url}')
        print(f'Pickup: {pickup_locode} at {pickup_date}')
        print(f'Return: {return_locode} at {return_date}')
        print(f'Driver Age: {driver_age}, Currency: {currency}')
        print(f'Agreement: {agreement_ref}')
        print()

        # Create availability criteria
        criteria = AvailabilityCriteria.make(
            pickup_locode,
            return_locode,
            datetime.fromisoformat(pickup_date.replace('Z', '+00:00')),
            datetime.fromisoformat(return_date.replace('Z', '+00:00')),
            driver_age,
            currency,
            [agreement_ref],
        )

        print('Searching availability...')
        print()

        # Search availability (streaming)
        chunk_count = 0
        total_offers = 0

        async for chunk in client.availability().search(criteria):
            chunk_count += 1
            status = chunk.get('status', 'PARTIAL')
            items = chunk.get('items', [])
            total_offers += len(items)

            print(f'[Chunk {chunk_count}] Status: {status}, Offers: {len(items)}')

            if items:
                # Show first offer as example
                first_offer = items[0]
                vehicle_class = first_offer.get('vehicle_class', 'N/A')
                make_model = first_offer.get('make_model', 'N/A')
                price = first_offer.get('total_price', 'N/A')
                offer_currency = first_offer.get('currency', currency)
                source_id = first_offer.get('source_id', 'N/A')
                print(f'  Example offer: {vehicle_class} - {make_model}')
                print(f'    Price: {offer_currency} {price}')
                print(f'    Source: {source_id}')

            if status == 'COMPLETE':
                print()
                print(f'✓ Search complete! Total chunks: {chunk_count}, Total offers: {total_offers}')
                break

        if chunk_count == 0:
            print('⚠ No availability chunks received')

    except Exception as error:
        print(f'❌ Error: {error}')
        if hasattr(error, 'status_code'):
            print(f'   Status Code: {error.status_code}')
        if hasattr(error, 'code'):
            print(f'   Error Code: {error.code}')
        sys.exit(1)


if __name__ == '__main__':
    import asyncio
    asyncio.run(test_availability())

